### PR TITLE
[GenericPoseGraph] Add RelativePoseCost2D.

### PR DESCRIPTION
**Pull request by pifon2a**
This is the implementation of the ceres::CostFunction that uses the new proto definition. It is tested using the Autodiff version.

### CODE REVIEW

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+      const proto::RelativePose2D::Parameters& parameters)
+      : translation_weight_(parameters.translation_weight()),
+        rotation_weight_(parameters.rotation_weight()),
+        first_T_second_(transform::ToRigid2(parameters.first_t_second())) {}
+
+  template <typename T>
+  bool operator()(const T* const start_pose, const T* const end_pose,
+                  T* e) const {
+    const std::array<T, 3> error = mapping::optimization::ScaleError(
+        mapping::optimization::ComputeUnscaledError(first_T_second_, start_pose,
```
Comment (mbokeloh) [inline cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc]: To me that's completely unclear why this is the same residual as the one in the implementation. Shouldn't we use this function as well to compute the residual in the first place?
Alternatively, could we use the residual only computation here? Maybe we can even extract a common pattern (I suppose that all cost functions could be tested this way). Let's maybe discuss this point.

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    analytical_cost_ = common::make_unique<
+        ceres::AutoDiffCostFunction<AutoDiffRelativePoseCost, kResidualsCount,
+                                    kPoseDimension, kPoseDimension>>(
+        new AutoDiffRelativePoseCost(parameters));
+    for (int i = 0; i < kParameterBlocksCount; ++i) {
+      jacobian_ptrs_[i] = jacobian_[i].data();
+    }
+  }
+
+  std::pair<const ResidualType&, const JacobianType&> EvaluateAnalyticalCost(
```
Comment (mbokeloh) [inline cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc]: Let's rename this to EvaluateRelativePoseCost2D to make it clear that this is the method we want to test. Please also rename analytical_cost_ accordingly.

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_POSE_GRAPH_CONSTRAINT_COST_FUNCTION_RELATIVE_POSE_COST_2D_H_
+#define CARTOGRAPHER_POSE_GRAPH_CONSTRAINT_COST_FUNCTION_RELATIVE_POSE_COST_2D_H_
+
+#include "cartographer/pose_graph/proto/cost_function.pb.h"
```
[the remark this answers is not in the document]
Comment (pifon2a) [reply cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h]: unfortunately, no :)

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+class RelativePoseCost2D
+    : public ceres::SizedCostFunction<3 /* number of residuals */,
+                                      3 /* size of first pose */,
+                                      3 /* size of second pose */> {
+ public:
+  explicit RelativePoseCost2D(
+      const proto::RelativePose2D::Parameters& Parameters);
+
+  proto::RelativePose2D::Parameters ToProto();
```
[the remark this answers is not in the document]
Comment (pifon2a) [reply cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h]: done

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+namespace cartographer {
+namespace pose_graph {
+
+class RelativePoseCost2D
+    : public ceres::SizedCostFunction<3 /* number of residuals */,
+                                      3 /* size of first pose */,
+                                      3 /* size of second pose */> {
+ public:
+  explicit RelativePoseCost2D(
+      const proto::RelativePose2D::Parameters& Parameters);
```
[the remark this answers is not in the document]
Comment (pifon2a) [reply cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h]: done

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  ResidualType residuals_;
+  JacobianType jacobian_;
+  std::array<double*, kParameterBlocksCount> jacobian_ptrs_;
+  std::unique_ptr<ceres::CostFunction> auto_diff_cost_;
+  std::unique_ptr<ceres::CostFunction> analytical_cost_;
+};
+
+::testing::Matcher<double> Near(double expected, double precision = 1e-05) {
```
[the remark this answers is not in the document]
Comment (pifon2a) [reply cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc]: done

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc (hunk 2 above)
Comment (pifon2a) [reply cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc]: done

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc (hunk 1 above)
Comment (pifon2a) [reply cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc]: done

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+  proto::RelativePose2D::Parameters ToProto();
+
+  bool Evaluate(double const* const* parameters, double* residuals,
```
[the remark this answers is not in the document]
Comment (pifon2a) [reply cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h]: done

File: cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.cc [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  *parameters.mutable_first_t_second() = transform::ToProto(first_T_second_);
+  return parameters;
+}
+
+bool RelativePoseCost2D::Evaluate(double const* const* parameters,
+                                  double* residuals, double** jacobians) const {
+  double const* start = parameters[0];
+  double const* end = parameters[1];
+
+  const double cos_start_rotation = cos(start[2]);
```
[the remark this answers is not in the document]
Comment (pifon2a) [reply cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.cc]: i think no

**Comment by pifon2a** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user pifon2a. Merge queue now has a length of 1.

**Pull request merged**